aws-lc-rs/src/cipher/streaming.rs

```diff
@@ -0,0 +1,840 @@
+    ///   the input in certain circumstances.
+    /// * Returns an error if the length of either `input` or `output` is larger
+    ///   than `i32::MAX`.
+    pub fn update<'a>(
+        &mut self,
+        input: &[u8],
+        output: &'a mut [u8],
+    ) -> Result<BufferUpdate<'a>, Unspecified> {
+        let mut outlen: i32 = output.len().try_into()?;
+        let inlen: i32 = input.len().try_into()?;
+        if 1 != unsafe {
+            EVP_DecryptUpdate(
+                self.cipher_ctx.as_mut_ptr(),
+                output.as_mut_ptr(),
+                &mut outlen,
+                input.as_ptr(),
+                inlen,
+            )
+        } {
+            return Err(Unspecified);
+        }
+        let outlen: usize = outlen.try_into()?;
+        Ok(BufferUpdate::new(output, outlen))
+    }
+
+    /// Finishes the decryption operation, writing the remaining plaintext to
+    /// `output`.
+    /// The number of bytes written to `output` can be up to the block length of
+    /// the cipher algorithm (e.g., 16 bytes for AES).
+    ///
+    /// # Errors
+    /// * May return an error if the `output` buffer is smaller than the algorithm's
+    ///   block length. Certain cipher modes (such as CTR) may allow the output buffer
+    ///   to only be large enough to fit the remaining plaintext.
+    /// * Returns an error if the length of `output` is larger than `i32::MAX`.
+    pub fn finish(self, output: &mut [u8]) -> Result<BufferUpdate, Unspecified> {
+        let mut outlen: i32 = output.len().try_into()?;
+        if 1 != unsafe { EVP_DecryptFinal_ex(*self.cipher_ctx, output.as_mut_ptr(), &mut outlen) } {
+            return Err(Unspecified);
+        }
+        let outlen: usize = outlen.try_into()?;
+        Ok(BufferUpdate::new(output, outlen))
+    }
+
+    /// Returns the cipher operating mode.
+    #[must_use]
+    pub fn mode(&self) -> OperatingMode {
+        self.mode
+    }
+
+    /// Returns the cipher algorithm
+    #[must_use]
+    pub fn algorithm(&self) -> &'static Algorithm {
+        self.algorithm
+    }
+
+    /// Constructs a `StreamingDecryptingKey` for decrypting using the CTR cipher mode.
+    /// The resulting plaintext will be the same length as the ciphertext.
+    ///
+    /// # Errors
+    /// Returns an error on an internal failure.
+    pub fn ctr(key: UnboundCipherKey, context: DecryptionContext) -> Result<Self, Unspecified> {
+        Self::new(key, OperatingMode::CTR, context)
+    }
+
+    /// Constructs a `StreamingDecryptingKey` for decrypting using the CBC cipher mode.
+    /// The resulting plaintext will be shorter than the ciphertext.
+    ///
+    /// # Errors
+    /// Returns an error on an internal failure.
+    pub fn cbc_pkcs7(
+        key: UnboundCipherKey,
+        context: DecryptionContext,
+    ) -> Result<Self, Unspecified> {
+        Self::new(key, OperatingMode::CBC, context)
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Can you add tests to the existing cipher FIPS module indicator tests to ensure we have coverage for this new type?